### PR TITLE
Circuit Execution Outputs

### PR DESCRIPTION
**Pull request by BrunoLiegiBastonLiegi**
As requested in #1020, three objects are implemented that substitute the old `CircuitResult` object. 
Depending on the different conditions, now the circuit execution returns:

1. no noise, no measurements --> `QuantumState`
2. no noise, measurements --> `CircuitResult`
3. noise, no measurements --> raise error if `density_matrix=False` else `QuantumState`
4. noise, measurements --> `CircuitResult` if `density_matrix=True` else `MeasurementOutcomes`

where the new `CircuitResult(QuantumState, MeasurementOutcomes)` object contains both the `QuantumState` and the `MeasurementOutcomes`.
`nshots` is set by default to `1000` now.
The probabilities and the state can be recovered as `QuantumState.probabilities()` and `QuantumState.state()`, the frequencies and the samples with `MeasurementOutcomes.frequencies()` and `MeasurementOutcomes.samples()`. `CircuitResult` gives access to all of them in the same way.

Currently the only case that has been left out is measurements with `collapse=True`. 
One possibility to handle them could be:
- only collapsing measurements present
      --> `QuantumState` if density_matrix=True
      --> raise Error if density_matrix=False
- both collapsing and standard measurements present
     --> `CircuitResult` if density_matrix=True
     --> `MeasurementOutcomes` if density_matrix=False

Furthermore, handling the `Circuit._final_state` attribute now is a bit problematic, as depending on the conditions different different things are obtained after execution.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

**Comment by BrunoLiegiBastonLiegi** [comment]
Okay this should be ready for review now. I am also working on adding a table to the docs to summarize the circuit execution outcomes depending on the settings.

File: tests/test_measurements.py
Comment (stavros11) [inline tests/test_measurements.py:479]:
```suggestion
def test_circuit_result_error(backend):
```
to follow the usual conventions. It's true that we are testing the class but capitals do not look good in function names.

File: tests/test_measurements_collapse.py
Comment (stavros11) [inline tests/test_measurements_collapse.py:9]:
```suggestion
```
dead code

File: tests/test_measurements_collapse.py
Comment (stavros11) [inline tests/test_measurements_collapse.py:32]:
```suggestion
```
same here assuming it is no longer valid.

File: tests/test_models_circuit_execution.py
Comment (stavros11) [inline tests/test_models_circuit_execution.py:10]: As in previous comments, prefer
```suggestion
    final_state = backend.execute_circuit(c).state()
```
outside the class.

**Review by stavros11 — APPROVED** [review]
Thank you for working on this @BrunoLiegiBastonLiegi. The new objects look good and the results are much cleaner now. I wrote some comments below, note that some of them are hidden because they are many. I still need to have a look in the documentation updates and a more detailed look in the test updates.

Other than that, it may be worth updating this with main and fixing the conflicts every now and then because conflicts will block the merge even if we approve. In principle you can fix once when we are ready to merge but is usually better to maintain this up-to-date with main.

**Comment by scarrazza** [comment]
@BrunoLiegiBastonLiegi thanks for this. Could you please also consider the possibility to store and load the numpy object that are required in order to reallocate the results classes? This is necessary for the future remote connection mechanism based on http requests.

File: src/qibo/states.py
Comment (BrunoLiegiBastonLiegi) [reply src/qibo/states.py:11]: `qubits` is actually used in the backend method `circuit_result_probabilities`:

```
    def circuit_result_probabilities(self, state, qubits=None):
        if qubits is None:
            qubits = state.qubits

        state_tensor = state._state
        if state.density_matrix:
            return self.calculate_probabilities_density_matrix(
                state_tensor, qubits, state.nqubits
            )
        else:
            return self.calculate_probabilities(state_tensor, qubits, state.nqubits)
```

maybe by setting  the default value for `calculate_probabilities` methods we could avoid this

File: src/qibo/states.py
Comment (BrunoLiegiBastonLiegi) [reply src/qibo/states.py:12]: `density_matrix` is needed for the `symbolic` method in `QuantumState`
`nqubits` and `qubits` are needed for the `circuit_result_probabilties` above
`_repeated_execution_probabilities` for the `_final_state` attribute of circuit and thus for when the circuit is `compile` -ed

File: src/qibo/backends/numpy.py
Comment (BrunoLiegiBastonLiegi) [reply src/qibo/backends/numpy.py:430]: In the end I decided to keep it, as I think the circuit compilation breaks otherwise. Still, it would probably be better to change the name to `_final_result` as now even a stateless object could be returned in principle.

File: tests/test_callbacks.py
Comment (BrunoLiegiBastonLiegi) [reply tests/test_callbacks.py:175]: yeah sorry about that, the thing is that I modified the `assert_allclose` only after correcting some of the tests manually, indeed I'll remove the `.state()` calls where unnecessary

File: src/qibo/states.py
Comment (stavros11) [reply src/qibo/states.py:11]: I see, thanks. Then a potential simplification is to remove `circuit_result_probabilities` from the backends and copy its content in the `QuantumState.probabilities`. I think this is the only place this is used anyway. Then the backends will be completely free from `circuit_result_*` methods and you can probably also drop `self.qubits` by doing
```py
if qubits is not None:
    # use all qubits
    qubits = tuple(range(self.nqubits))
```
inside this method.

File: src/qibo/backends/numpy.py
Comment (stavros11) [reply src/qibo/backends/numpy.py]: Given that we may drop this completely in the future, I wouldn't worry much about the name. However we could open an issue to remember checking if this really needed and if yes maybe update the name in the future.

File: src/qibo/states.py
Comment (stavros11) [reply src/qibo/states.py:12]: Thanks, I think I meant only the `_repeated_execution_probabilities`. Are you sure it is related to the `_final_state`/`compile`? I don't think these are releated to repeated execution.

File: src/qibo/backends/numpy.py
Comment (stavros11) [inline src/qibo/backends/numpy.py:531]: You can set the `_final_state` and return only once at the end of this function instead of every if case. This will save some repeated lines of code.

File: src/qibo/backends/numpy.py
Comment (stavros11) [inline src/qibo/backends/numpy.py:535]: Regarding the `QuantumState._repeated_execution_probabilities` (comment above), I think this is the only place they are used. However if I follow the code properly I think it is just doing a trivial caching. You are calculating the probabilities from the `final_result` and then you are setting them back to the same object. If you call `final_result` again after that you will get the values cached in `_repeated_execution_probabilities` but these will be the same with just redoing the calculation. If that's correct, it should be safe to drop this variable completely.

If you need caching, you can handle it internally in the `QuantumState` object, or use [standard Python methods.](https://docs.python.org/3/library/functools.html#functools.cache). However I don't think that's needed in our case.

File: src/qibo/backends/numpy.py
Comment (stavros11) [inline src/qibo/backends/numpy.py:544]: Also here you are assigning the `_repeated_execution_probabilities` attribute to a `MeasurementOutcomes` object, but this does not exist in the original class definition. Python allows you to set attributes to instantiated objects, even if they do not exist in their definition but it's not a very good practice because then it's harder to track what attributes each object has. Pylint usually complains for this, I am not sure why it is passing now.

If the `_repeated_execution_probabilities` is not used we can remove, otherwise we should define it as `None` in the `MeasurementOutcomes.__init__`.

File: tests/test_cirq.py
Comment (stavros11) [inline tests/test_cirq.py:62]: This should be replaced with
```suggestion
            final_state = backend.execute_circuit(c, np.copy(initial_state)).state()
```
to avoid using the "private" attribute. Same for three other instances in this file.

File: tests/test_gates_density_matrix.py
Comment (stavros11) [inline tests/test_gates_density_matrix.py:168]: Same as above in this file.

File: tests/test_hamiltonians_terms.py
Comment (stavros11) [inline tests/test_hamiltonians_terms.py:53]: Same here.

File: tests/test_hamiltonians_trotter.py
Comment (stavros11) [inline tests/test_hamiltonians_trotter.py:136]: Same here.

File: tests/test_measurements.py
Comment (stavros11) [inline tests/test_measurements.py:265]: Same here.

File: tests/test_measurements_collapse.py
Comment (stavros11) [inline tests/test_measurements_collapse.py:115]: For this and other tests that are skipped, we should decide whether:
* we keep as skipped but write a more explicit reason, for example: skipped because we no longer support repeated execution for density matrices (or something like that), or
* remove completely.

Either solution is fine with me.

File: tests/test_measurements_collapse.py
Comment (stavros11) [inline tests/test_measurements_collapse.py:165]:
```suggestion
```

File: tests/test_models_circuit_noise.py
Comment (stavros11) [inline tests/test_models_circuit_noise.py:23]: Same as above
```suggestion
    final_rho = backend.execute_circuit(c)
```

I think in the cases where you use `backend.assert_allclose`, you don't need to call `.state()` explicitly because this is done automatically in the backend method.

File: tests/test_models_circuit_parametrized.py
Comment (stavros11) [inline tests/test_models_circuit_parametrized.py:208]: Same as above.

File: tests/test_models_circuit_qasm_cirq.py
Comment (stavros11) [inline tests/test_models_circuit_qasm_cirq.py:31]: Same as above.

File: src/qibo/result.py
Comment (stavros11) [inline src/qibo/result.py:114]: Looking also at the qibolab PR, it could be useful to have a `probabilities` method in the `MeasurementOutcomes` too. This would return `self._probs` if not `None`, otherwise it would calculate the probabilities as frequencies / nshots.

Then in `CircuitResult.probabilities` you just need to make sure that you use the `QuantumState.probabilities` instead of the `MeasurementOutcome.probabilities`.

**Review by stavros11 — COMMENTED** [review]
Thanks for the updates @BrunoLiegiBastonLiegi, some additional comments. Also, I see an empty file called `test_result.py~`.

File: src/qibo/result.py
Comment (scarrazza) [inline src/qibo/result.py:92]: Please rename this to `to_dict`.

**Review by scarrazza — APPROVED** [review]

**Review by stavros11 — APPROVED** [review]
I had a quick look at the docs as I have not done this before and it looks good to me. Feel free to merge this and open an issue about the probabilities or anything else we could potentially address in the future.

I will test the qibolab part later and if there is any update required here we can fix in a new PR.

**Pull request merged**
